### PR TITLE
`Programming exercises`: Fix an issue in which long manual feedback is not correctly displayed

### LINKED ISSUE #9434: General manual feedback not fully shown in feedback view
### Describe the bug

After professor uploaded feedback to exercise, while trying to extend it, it's still not being shown properly, ending by [...].
Photo is attached.
The issue is same for tablet and phone version, even while using "Desktop site" functionality

### To Reproduce

1. Go to 'exercise
2. Click on last review and then on 'feedback'
3. Scroll down to 'end of feedback'
4. Not be able to extend feedback any more even though it has some more text
5. See error

### Expected behavior

After expanding feedback I expected to see the whole text in phone version

### Screenshots

Screenshot_20241007_150610_Chrome

### Which version of Artemis are you seeing the problem on?

7.5.6

### What browsers are you seeing the problem on?

Chrome, Safari

### Additional context

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by az108**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **Gitlab** and **Jenkins**.

### Motivation and Context

Fixes #9434
This is a fix which adresses following issue where long feedback was not beeing loaded properly from the server.
This is most probably either caused by the result beeing damaged or not saved properly or the feedback beeing overwritten. To tackle the second mentioned potential reason following was done (described in Description).

Additionally another bug was fixed where instructors were not able to edit long feedback in manual assessements.

### Description

I refactored the method which handles saving the long feedback when its created and when a long feedback is edited then the whole feedback is loaded to avoid that instructors override the long feedback with the short feedback which is loaded per default. The BaseAPIHTTPService on the client side was adjusted to also accept text responses from the server, making the rest call look very easy to understandable. 

Note: The saving of manual assesements for programming exercise needs further refactoring but this is not part of this PR!

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise 

1. Log in to Artemis as Instructor
2. Create a Programming Exercise
3. Log in as Student and participate in the Programming Exercise
4. Log in as Instructor and set the Due Date to now and set Manual Assesement 
5. Navigate to the Programming Exercise and press the View button in the TutorActions tab
6. Press the Assesement Dashboard button
7. Assess the Student Sumbission by creating a new Feedback with over 1000 characters
8. Save this and reload the page
9. Verify that the feedback box now displays the whole long feedback and no shortend version
10. Try to edit the long feedback and save again
11. Verify that the long feedback was changed
12. Do this for Modelling and Quiz and Text Exercises aswell

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review

- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Performance Tests
- [x] Test 1
- [x] Test 2

### Screenshots

Before:
image
After:
image

## Release Notes

- **New Features**
	- Introduced handling for long feedback texts across various assessment services.
	- Added conditional rendering for feedback components based on the presence of valid assessment results.
	- Enhanced `FeedbackService` to fetch long feedback text dynamically.
	- Implemented new input properties for components to receive result identifiers, improving feedback detail processing.

- **Bug Fixes**
	- Enhanced feedback management to prevent duplicates during assessments.

- **Documentation**
	- Updated comments for clarity in the `BaseApiHttpService`.

- **Tests**
	- Improved test coverage for `FeedbackService` with Angular's testing utilities and HTTP request mocking.

### CODE REVIEW

**Review by sachmii — APPROVED** [review]
Tested on TS5. Feedback can be uncollapsed and doesn't show up as [...] when longer than 1000 character.

Bildschirmfoto 2024-10-24 um 23 35 11
Bildschirmfoto 2024-10-24 um 23 34 46

**Review by sachmii — DISMISSED** [review]
re-approve

**Review by Malekos74 — APPROVED** [review]
### Server tested on: Server 2
Works as expected for all types of exercices
**Problems:**
- Nothing to report
**Minor notice:**
- No notices
**Requests:**
- Maybe have the feedback automatically unroll to be the size of the text

**Review by HawKhiem — APPROVED** [review]
Tested on TS4. Eveything works as described. One small thing: I had to scroll to extend the feedback box. Is there any way to improve this, perhaps like an extend button or sort

https://github.com/user-attachments/assets/47e80c45-bb8c-4871-8c57-9b30056463e5

File: src/main/webapp/app/exercises/modeling/manage/example-modeling/example-modeling-submission.component.html
Comment (florian-glombik) [inline src/main/webapp/app/exercises/modeling/manage/example-modeling/example-modeling-submission.component.html:192]: Is the non null assertion really needed here?
```suggestion
                                [resultId]="result.id"
```

File: src/main/webapp/app/assessment/unreferenced-feedback-detail/unreferenced-feedback-detail.component.ts
Comment (florian-glombik) [inline src/main/webapp/app/assessment/unreferenced-feedback-detail/unreferenced-feedback-detail.component.ts:16]: Please use a signal here if possible and mark the resulted as required if possible

File: src/main/webapp/app/assessment/unreferenced-feedback-detail/unreferenced-feedback-detail.component.ts
Comment (florian-glombik) [inline src/main/webapp/app/assessment/unreferenced-feedback-detail/unreferenced-feedback-detail.component.ts:57]: Why are we not directly sending the longFeedbackText with the feedback? 
I think it would be a better approach to fetch the longFeedbackText directly from the server when fetching the feedback.

This has the potential to slightly increase the performance load on the server 😅

File: src/main/java/de/tum/cit/aet/artemis/assessment/service/AssessmentService.java
Comment (florian-glombik) [inline src/main/java/de/tum/cit/aet/artemis/assessment/service/AssessmentService.java:290]: I think it would be better to use javadoc here `{@link resultRepository.save}` to ensure that the comment is updated if something is refactored / renamed

Comment (florian-glombik) [reply src/main/java/de/tum/cit/aet/artemis/assessment/service/AssessmentService.java:290]: Is this the right place for the fix then? It feels a little awkward that something has to be deleted for save to work properly - this is not how I would expect save to work 😅

File: src/main/java/de/tum/cit/aet/artemis/assessment/service/ResultService.java
Comment (florian-glombik) [inline src/main/java/de/tum/cit/aet/artemis/assessment/service/ResultService.java:517]: Why do we need to do this?

File: src/main/java/de/tum/cit/aet/artemis/assessment/service/ResultService.java
Comment (florian-glombik) [inline src/main/java/de/tum/cit/aet/artemis/assessment/service/ResultService.java:539]: Could you please describe the actual problem that shall be avoided aswell?

File: src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingAssessmentService.java
Comment (florian-glombik) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingAssessmentService.java:97]: This seems to be a code duplicate with the same 4 lines in `src/main/java/de/tum/cit/aet/artemis/assessment/service/ResultService.java` - could you please generalize it to a method?

**Review by florian-glombik — CHANGES_REQUESTED** [review]
I have left some comments and some questions.

I do not understand why we have the separation between feedback and long feedback. Was that an active decision to save payload? 
If that was the case, then it might make sense to fetch the long feedback individually - however, if I didn't misunderstand the code, we are now always fetching the long feedback in the onInit, which means that we do not save any payload at all and just send more requests? Is this correct?

If I am not mistaken we should at least fix this by only loading the long feedback when it is uncollapsed, otherwise I do not see any benefit of the separation of the feedback and the long feedback text (I might miss something obvious here)

File: src/main/java/de/tum/cit/aet/artemis/assessment/service/AssessmentService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/assessment/service/AssessmentService.java:291]: Do we do this deletion here, because we only want to associate the long feedbacks with the newest result and not any old result?
Do we really want to break the old results so students can no longer look at them?

File: src/main/java/de/tum/cit/aet/artemis/assessment/service/ResultService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/assessment/service/ResultService.java:655]:
```suggestion
        longFeedbackRepository.deleteAllByFeedbackIds(feedbackList.stream().map(DomainObject::getId).toList());
```
Otherwise this might result in many data base calls

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/cit/aet/artemis/assessment/service/AssessmentService.java
Comment (az108) [reply src/main/java/de/tum/cit/aet/artemis/assessment/service/AssessmentService.java:291]: Yes, we need this to avoid a duplicate entry error in the LongFeedbackRepository when saving a result in the ResultRepository. This approach doesn’t break the previous results, as it only applies when saving manual feedback after the due date. This means we aren’t working with multiple results; we’re only modifying a single result, which we overwrite each time.
image

File: src/main/webapp/app/assessment/unreferenced-feedback-detail/unreferenced-feedback-detail.component.ts
Comment (az108) [reply src/main/webapp/app/assessment/unreferenced-feedback-detail/unreferenced-feedback-detail.component.ts:57]: its actually pretty rare that long feedback exists, i do not really see the point of sending a request everytime. we should only send it when a long feedback also exists

File: src/main/java/de/tum/cit/aet/artemis/assessment/service/AssessmentService.java
Comment (az108) [reply src/main/java/de/tum/cit/aet/artemis/assessment/service/AssessmentService.java:290]: the deletion is also weird for me but saving results does not work otherwise as resultRepository.save for some reason also saves the longfeedbacktext in the longfeedbacktextrepository. And because we only override the latest result with a manual correction i do not really see a issue deleting the existing longfeedback from that seperate repository

File: src/main/java/de/tum/cit/aet/artemis/assessment/service/ResultService.java
Comment (az108) [reply src/main/java/de/tum/cit/aet/artemis/assessment/service/ResultService.java:517]: this was code which i only refactored and did not write myself. apparently there was a issue related to this https://github.com/ls1intum/Artemis/pull/7792

**Review by sachmii — APPROVED** [review]

File: src/main/java/de/tum/cit/aet/artemis/assessment/service/ResultService.java
Comment (florian-glombik) [reply src/main/java/de/tum/cit/aet/artemis/assessment/service/ResultService.java:655]: At least in the diff that is displayed to me right now this does not seem to be addressed - maybe my diff is outdated and doesn't reload properly? 🤔

**Review by florian-glombik — COMMENTED** [review]
Thanks for incorporating the change requests, the code looks good for me now except that the [change request of Johannes](https://github.com/ls1intum/Artemis/pull/9562#discussion_r1817770295) does not seem to be addressed, could you double check that the suggestion was applied? (Which does not seem to be the case for me)

File: src/main/java/de/tum/cit/aet/artemis/assessment/service/ResultService.java
Comment (az108) [reply src/main/java/de/tum/cit/aet/artemis/assessment/service/ResultService.java:655]: The approach actually does not work. I could not answer why

**Review by flbrgit — COMMENTED** [review]
I tested on TS2 with a programming exercise and a feedback text of 1823 characters. I followed the steps provided, but still got the shortened version in the feedback-tab after saving:
Screenshot 2024-10-27 105846
However, after submitting the full long text, the student feedback was not shortened and worked as expected:
Screenshot 2024-10-27 111602

**Comment by az108** [comment]
@Malekos74 Making the size of the box adjustable to the feedback text would be part of a refactor and not bugfix. will be considered as i agree that its weird in its current situation.

**Comment by az108** [comment]
@flbrgit I hope the issue is resolved now :)

**Review by florian-glombik — DISMISSED** [review]
Approve code

**Review by Malekos74 — DISMISSED** [review]
Code approve 🫡🫡

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by florian-glombik — APPROVED** [review]
Re-approve code after merging develop

**Review by sachmii — APPROVED** [review]
Re-approve

**Review by Malekos74 — APPROVED** [review]
Merge conflict reapprove

**Review by SindiBuklaji — APPROVED** [review]
Tested on TS5. Works as described 👍

**Pull request merged**
